File: vandalizer-next/backend/app/routers/knowledge.py
"""Knowledge Base API routes."""

from fastapi import APIRouter, Depends, HTTPException

from app.dependencies import get_current_user
from app.models.user import User
from app.services import group_service
from app.schemas.knowledge import (
    AddDocumentsRequest,
    AddUrlsRequest,
    CreateKBRequest,
    KBDetailResponse,
    KBResponse,
    KBSourceResponse,
    KBStatusResponse,
    UpdateKBRequest,
)
from app.services import knowledge_service as svc

router = APIRouter()


def _kb_response(kb) -> KBResponse:
    return KBResponse(
        uuid=kb.uuid,
        title=kb.title,
        description=kb.description or "",
        status=kb.status,
        shared_with_team=kb.shared_with_team,
        verified=kb.verified,
        group_ids=kb.group_ids,
        total_sources=kb.total_sources,
        sources_ready=kb.sources_ready,
        sources_failed=kb.sources_failed,
        total_chunks=kb.total_chunks,
        created_at=kb.created_at.isoformat() if kb.created_at else None,
        updated_at=kb.updated_at.isoformat() if kb.updated_at else None,
    )


def _source_response(s) -> KBSourceResponse:
    return KBSourceResponse(
        uuid=s.uuid,
        source_type=s.source_type,
        document_uuid=s.document_uuid,
        url=s.url,
        url_title=s.url_title or "",
        status=s.status,
        error_message=s.error_message or "",
        chunk_count=s.chunk_count,
        created_at=s.created_at.isoformat() if s.created_at else None,
    )


@router.get("/list", response_model=list[KBResponse])
async def list_knowledge_bases(user: User = Depends(get_current_user)):
    team_id = str(user.current_team) if user.current_team else None
    user_group_uuids = await group_service.get_user_group_uuids(user.user_id)
    kbs = await svc.list_knowledge_bases(
        user.user_id, team_id=team_id, user_group_uuids=user_group_uuids,
    )
    return [_kb_response(kb) for kb in kbs]


@router.post("/create", response_model=KBResponse)
async def create_knowledge_base(req: CreateKBRequest, user: User = Depends(get_current_user)):
    if not req.title.strip():
        raise HTTPException(status_code=400, detail="Title is required")
    team_id = str(user.current_team) if user.current_team else None
    kb = await svc.create_knowledge_base(
        title=req.title, user_id=user.user_id,
        team_id=team_id, description=req.description,
    )
    return _kb_response(kb)


@router.get("/{uuid}", response_model=KBDetailResponse)
async def get_knowledge_base(uuid: str, user: User = Depends(get_current_user)):
    kb = await svc.get_knowledge_base(uuid, user.user_id)
    if not kb:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    sources = await svc.get_kb_sources(kb.uuid)
    return KBDetailResponse(
        **_kb_response(kb).model_dump(),
        sources=[_source_response(s) for s in sources],
    )


@router.post("/{uuid}/update")
async def update_knowledge_base(uuid: str, req: UpdateKBRequest, user: User = Depends(get_current_user)):
    kb = await svc.update_knowledge_base(
        uuid, user.user_id, title=req.title, description=req.description,
        shared_with_team=req.shared_with_team, group_ids=req.group_ids,
    )
    if not kb:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    return {"ok": True}


@router.post("/{uuid}/share")
async def share_knowledge_base(uuid: str, user: User = Depends(get_current_user)):
    kb = await svc.share_with_team(uuid, user.user_id)
    if not kb:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    return {"ok": True, "shared_with_team": kb.shared_with_team}


@router.delete("/{uuid}")
async def delete_knowledge_base(uuid: str, user: User = Depends(get_current_user)):
    ok = await svc.delete_knowledge_base(uuid, user.user_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    return {"ok": True}


@router.post("/{uuid}/add_documents")
async def add_documents(uuid: str, req: AddDocumentsRequest, user: User = Depends(get_current_user)):
    kb = await svc.get_knowledge_base(uuid, user.user_id)
    if not kb:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    if not req.document_uuids:
        raise HTTPException(status_code=400, detail="No documents provided")
    kb.status = "building"
    await kb.save()
    added = await svc.add_documents(kb, req.document_uuids)
    return {"ok": True, "added": added}


@router.post("/{uuid}/add_urls")
async def add_urls(uuid: str, req: AddUrlsRequest, user: User = Depends(get_current_user)):
    kb = await svc.get_knowledge_base(uuid, user.user_id)
    if not kb:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    if not req.urls:
        raise HTTPException(status_code=400, detail="No URLs provided")
    kb.status = "building"
    await kb.save()
    added = await svc.add_urls(
        kb, req.urls,
        crawl_enabled=req.crawl_enabled,
        max_crawl_pages=req.max_crawl_pages,
        allowed_domains=req.allowed_domains,
    )
    return {"ok": True, "added": added}


@router.delete("/{uuid}/source/{source_uuid}")
async def remove_source(uuid: str, source_uuid: str, user: User = Depends(get_current_user)):
    kb = await svc.get_knowledge_base(uuid, user.user_id)
    if not kb:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    ok = await svc.remove_source(kb, source_uuid)
    if not ok:
        raise HTTPException(status_code=404, detail="Source not found")
    return {"ok": True}


@router.get("/{uuid}/status", response_model=KBStatusResponse)
async def get_status(uuid: str, user: User = Depends(get_current_user)):
    kb = await svc.get_knowledge_base(uuid, user.user_id)
    if not kb:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    sources = await svc.get_kb_sources(kb.uuid)
    return KBStatusResponse(
        uuid=kb.uuid,
        status=kb.status,
        total_sources=kb.total_sources,
        sources_ready=kb.sources_ready,
        sources_failed=kb.sources_failed,
        total_chunks=kb.total_chunks,
        sources=[
            {"uuid": s.uuid, "status": s.status, "error_message": s.error_message or "", "chunk_count": s.chunk_count}
            for s in sources
        ],
    )

File: frontend/src/api/knowledge.ts
import { apiFetch } from './client'
import type { KnowledgeBase, KnowledgeBaseDetail } from '../types/knowledge'

export function listKnowledgeBases() {
  return apiFetch<KnowledgeBase[]>('/api/knowledge/list')
}

export function createKnowledgeBase(title: string, description?: string) {
  return apiFetch<KnowledgeBase>('/api/knowledge/create', {
    method: 'POST',
    body: JSON.stringify({ title, description }),
  })
}

export function getKnowledgeBase(uuid: string) {
  return apiFetch<KnowledgeBaseDetail>(`/api/knowledge/${uuid}`)
}

export function updateKnowledgeBase(uuid: string, data: { title?: string; description?: string }) {
  return apiFetch<{ ok: boolean }>(`/api/knowledge/${uuid}/update`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function deleteKnowledgeBase(uuid: string) {
  return apiFetch<{ ok: boolean }>(`/api/knowledge/${uuid}`, { method: 'DELETE' })
}

export function addDocumentsToKB(uuid: string, documentUuids: string[]) {
  return apiFetch<{ ok: boolean; added: number }>(`/api/knowledge/${uuid}/add_documents`, {
    method: 'POST',
    body: JSON.stringify({ document_uuids: documentUuids }),
  })
}

export function addUrlsToKB(
  uuid: string,
  urls: string[],
  crawlEnabled = false,
  maxCrawlPages = 5,
  allowedDomains = '',
) {
  return apiFetch<{ ok: boolean; added: number }>(`/api/knowledge/${uuid}/add_urls`, {
    method: 'POST',
    body: JSON.stringify({
      urls,
      crawl_enabled: crawlEnabled,
      max_crawl_pages: maxCrawlPages,
      allowed_domains: allowedDomains,
    }),
  })
}

export function removeKBSource(uuid: string, sourceUuid: string) {
  return apiFetch<{ ok: boolean }>(`/api/knowledge/${uuid}/source/${sourceUuid}`, {
    method: 'DELETE',
  })
}

export function shareKnowledgeBase(uuid: string) {
  return apiFetch<{ ok: boolean; shared_with_team: boolean }>(`/api/knowledge/${uuid}/share`, {
    method: 'POST',
  })
}

export function submitKBForVerification(kbUuid: string, data: {
  summary?: string
  description?: string
  category?: string
}) {
  return apiFetch<Record<string, unknown>>('/api/verify/submit', {
    method: 'POST',
    body: JSON.stringify({
      item_kind: 'knowledge_base',
      item_id: kbUuid,
      ...data,
    }),
  })
}

export function setKBOrganizations(uuid: string, organizationIds: string[]) {
  return apiFetch<{ ok: boolean }>(`/api/knowledge/${uuid}/update`, {
    method: 'POST',
    body: JSON.stringify({ organization_ids: organizationIds }),
  })
}

export function getKBStatus(uuid: string) {
  return apiFetch<{
    uuid: string
    status: string
    total_sources: number
    sources_ready: number
    sources_failed: number
    total_chunks: number
    sources: { uuid: string; status: string; error_message: string; chunk_count: number }[]
  }>(`/api/knowledge/${uuid}/status`)
}

// Validation

export function runKBValidation(uuid: string) {
  return apiFetch<Record<string, unknown>>(`/api/knowledge/${uuid}/validate`, {
    method: 'POST',
  })
}

export function getKBSourceHealth(uuid: string) {
  return apiFetch<{
    total: number
    healthy: number
    unhealthy: number
    ratio: number
    details: { uuid: string; source_type: string; name: string; status: string; error?: string }[]
  }>(`/api/knowledge/${uuid}/source-health`)
}

export function getKBQuality(uuid: string) {
  return apiFetch<{
    history: Record<string, unknown>[]
    contract: Record<string, unknown>
  }>(`/api/knowledge/${uuid}/quality`)
}

// Test queries

export function listKBTestQueries(uuid: string) {
  return apiFetch<{
    test_queries: {
      uuid: string
      query: string
      expected_source_labels: string[]
      expected_answer_contains: string | null
      created_at: string | null
    }[]
  }>(`/api/knowledge/${uuid}/test-queries`)
}

export function createKBTestQuery(uuid: string, data: {
  query: string
  expected_source_labels?: string[]
  expected_answer_contains?: string
}) {
  return apiFetch<Record<string, unknown>>(`/api/knowledge/${uuid}/test-queries`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function deleteKBTestQuery(uuid: string, queryUuid: string) {
  return apiFetch<{ ok: boolean }>(`/api/knowledge/${uuid}/test-queries/${queryUuid}`, {
    method: 'DELETE',
  })
}

// Clone

export function cloneKnowledgeBase(uuid: string, title?: string) {
  return apiFetch<Record<string, unknown>>(`/api/knowledge/${uuid}/clone`, {
    method: 'POST',
    body: JSON.stringify({ title }),
  })
}

// Suggestions

export function submitKBSuggestion(uuid: string, data: {
  suggestion_type: string
  url?: string
  document_uuid?: string
  note?: string
}) {
  return apiFetch<Record<string, unknown>>(`/api/knowledge/${uuid}/suggestions`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function listKBSuggestions(uuid: string) {
  return apiFetch<{
    suggestions: {
      uuid: string
      suggestion_type: string
      url: string | null
      document_uuid: string | null
      note: string | null
      status: string
      suggested_by_name: string | null
      suggested_by_user_id: string
      reviewed_by_user_id: string | null
      reviewed_at: string | null
      created_at: string | null
    }[]
  }>(`/api/knowledge/${uuid}/suggestions`)
}

export function reviewKBSuggestion(kbUuid: string, suggestionUuid: string, accept: boolean) {
  return apiFetch<Record<string, unknown>>(`/api/knowledge/${kbUuid}/suggestions/${suggestionUuid}`, {
    method: 'PATCH',
    body: JSON.stringify({ accept }),
  })
}
